juju/juju
# Uniter log

## Description of change

The following change adds the ability to log out from the charm in restricted
hook contexts that would normally prevent such. Feedback through the logs
could aid users of the charms in situations where things aren't been wired up
correctly.

Note: the prefix is dropped because there is no `HookRelation`!

> Why is this change needed?

Allow developers of charms the ability to debug the collect-metrics when writing
the charms. It could also provide useful feedback for users trying to watch the 
debug-log (for example).

## QA steps

> How do we verify that the change works?

Add a log to the collect metrics hook of a charm and it should now be available.

To fully test the following PR changes, you can use the following setup guide: 
Download the following file - [log-diff.txt](https://github.com/juju/juju/files/2098255/log-diff.txt)

```sh
mkdir $HOME/tmp
cd $HOME/tmp
charm pull cs:~cmars/mattermost
mkdir mattermost/config
echo "{}" > mattermost/config/config.json
cd mattermost
patch -p1 < /PATH/TO/FILE/log-diff.txt
cd ..
tar -zcvf mattermost.tar.gz mattermost
cd $GOPATH/src/github.com/juju/juju
make go-build go-install
juju bootstrap --build-agent=true localhost lxd-test
juju add-model lxd-test
juju deploy $HOME/tmp/mattermost --resource bdist=$HOME/tmp/mattermost.tar.gz
juju deploy postgresql
juju add-relation mattermost postgresql:db
juju debug-log
```

## Documentation changes

> Does it affect current user workflow? CLI? API?

Users/Developers of the charms.

## Bug reference

> Does this change fix a bug? Please add a link to it.

https://bugs.launchpad.net/juju/+bug/1763407
